packages/js/client-config-builder/src/BaseClientConfigBuilder.ts

```diff
@@ -0,0 +1,348 @@
+    const envUri = Uri.from(uri);
+
+    const idx = this.config.envs.findIndex((x) => Uri.equals(x.uri, envUri));
+
+    if (idx > -1) {
+      this.config.envs[idx].env = env;
+    } else {
+      this.config.envs.push({
+        uri: envUri,
+        env: env,
+      });
+    }
+
+    return this;
+  }
+
+  addInterfaceImplementation(
+    interfaceUri: Uri | string,
+    implementationUri: Uri | string
+  ): ClientConfigBuilder {
+    const interfaceUriSanitized = Uri.from(interfaceUri);
+    const implementationUriSanitized = Uri.from(implementationUri);
+
+    const existingInterface = this.config.interfaces.find((x) =>
+      Uri.equals(x.interface, interfaceUriSanitized)
+    );
+
+    if (existingInterface) {
+      if (
+        !existingInterface.implementations.some((x) =>
+          Uri.equals(x, implementationUriSanitized)
+        )
+      ) {
+        existingInterface.implementations.push(implementationUriSanitized);
+      }
+    } else {
+      this.config.interfaces.push({
+        interface: interfaceUriSanitized,
+        implementations: [implementationUriSanitized],
+      });
+    }
+
+    return this;
+  }
+
+  addInterfaceImplementations(
+    interfaceUri: Uri | string,
+    implementationUris: Array<Uri | string>
+  ): ClientConfigBuilder {
+    const interfaceUriSanitized = Uri.from(interfaceUri);
+    const implementationUrisSanitized = implementationUris.map(Uri.from);
+
+    const existingInterface = this.config.interfaces.find((x) =>
+      Uri.equals(x.interface, interfaceUriSanitized)
+    );
+
+    if (existingInterface) {
+      for (const implUri of implementationUrisSanitized) {
+        if (
+          !existingInterface.implementations.some((x) => Uri.equals(x, implUri))
+        ) {
+          existingInterface.implementations.push(implUri);
+        }
+      }
+    } else {
+      this.config.interfaces.push({
+        interface: interfaceUriSanitized,
+        implementations: implementationUrisSanitized,
+      });
+    }
+
+    return this;
+  }
+
+  removeInterfaceImplementation(
+    interfaceUri: Uri | string,
+    implementationUri: Uri | string
+  ): ClientConfigBuilder {
+    const interfaceUriSanitized = Uri.from(interfaceUri);
+    const implementationUriSanitized = Uri.from(implementationUri);
+
+    const existingInterface = this.config.interfaces.find((x) =>
+      Uri.equals(x.interface, interfaceUriSanitized)
+    );
+
+    if (existingInterface) {
+      const idx = existingInterface.implementations.findIndex((x) =>
+        Uri.equals(x, implementationUriSanitized)
+      );
+
+      if (idx > -1) {
+        existingInterface.implementations.splice(idx, 1);
+      }
+
+      if (existingInterface.implementations.length === 0) {
+        this.config.interfaces.splice(
+          this.config.interfaces.indexOf(existingInterface),
+          1
+        );
+      }
+    }
+
+    return this;
+  }
+
+  addRedirect(from: Uri | string, to: Uri | string): ClientConfigBuilder {
```

Comment:
What do you think about having redirects as a `map` instead of a `list` of `from-to`?

Reply:
We talked about making envs a map, but yea, redirects could also make sense. I think the same applies to wrappers and packages.
I'll look into the best way to do that in a follow-up PR